truffle/src/main/java/org/jruby/truffle/parser/MethodTranslator.java

```diff
@@ -449,4 +469,33 @@ private static String indentation(String line) {
         return "";
     }
 
+    /*
+     * The following methods allow us to save and restore enough of
+     * the current state of the Translator to allow lazy parsing. When
+     * the lazy parsing is actually performed, the state is restored
+     * to what it would have been if the method had been parsed
+     * eagerly.
+     */
+    public TranslatorState getCurrentState() {
+        return new TranslatorState(getEnvironment().unsafeGetLexicalScope(), getEnvironment().isDynamicConstantLookup(), new ArrayDeque<>(parentSourceSection));
```

Comment:
I don't recall the exact issue, but I had to do it in BodyTranslator a while ago (see f9df82cff1b48fb56545637de3bd1e605d19c11a). I think there was a problem with nested visit calls that each modified the `parentSourceSection`, which is just a field on the translator. Perhaps more locally scoped variables could work, but the stack approach made sense to me.